sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigtable/BigtableIO.java

```diff
@@ -1247,33 +1315,109 @@ public void startBundle(StartBundleContext c) throws IOException {
         bigtableWriter = serviceEntry.getService().openForWriting(writeOptions);
       }
+
+      badRecords = new HashSet<>();
+      unbatchedWrites = new HashSet<>();
     }
 
     @ProcessElement
     public void processElement(ProcessContext c, BoundedWindow window) throws Exception {
       checkForFailures();
       KV<ByteString, Iterable<Mutation>> record = c.element();
-      bigtableWriter
-          .writeRecord(record)
-          .whenComplete(
-              (mutationResult, exception) -> {
-                if (exception != null) {
-                  failures.add(new BigtableWriteException(record, exception));
-                }
-              });
+      bigtableWriter.writeRecord(record).whenComplete(handleMutationException(record, window));
       ++recordsWritten;
       seenWindows.compute(window, (key, count) -> (count != null ? count : 0) + 1);
     }
 
+    public BiConsumer<MutateRowResponse, Throwable> handleMutationException(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      return (MutateRowResponse result, Throwable exception) -> {
+        if (exception != null) {
+          if (exception instanceof NotFoundException
+              && !((NotFoundException) exception).isRetryable()) {
+            // This case, of being an NotFoundException and not retryable,
+            // indicates an issue with the data. In general, the likely cause is the user trying to
+            // write to a column family that doesn't exist. The frequency of this is dependent on
+            // how column families are determined upstream of the io. We can't know if this record
+            // failed, or if it was another record in the batch. Thus, we retry each record
+            // individually.
+            retryIndividualRecord(record, window);
+          } else {
+            failures.add(new BigtableWriteException(record, exception));
+          }
+        }
+      };
+    }
+
+    public void retryIndividualRecord(
+        KV<ByteString, Iterable<Mutation>> record, BoundedWindow window) {
+      try {
+        unbatchedWrites.add(
+            bigtableWriter
+                .writeSingleRecord(record)
+                .whenComplete(
+                    (singleMutationResult, singleException) -> {
+                      if (singleException != null) {
+                        if (singleException instanceof NotFoundException
+                            && !((NotFoundException) singleException).isRetryable()) {
+                          // if we get another NotFoundException, we know this is the bad record.
+                          badRecords.add(
+                              KV.of(new BigtableWriteException(record, singleException), window));
+                        } else {
+                          failures.add(new BigtableWriteException(record, singleException));
+                        }
+                      }
+                    }));
```

Comment:
You seem to have result processing split on 2 different threads:
- this sets up handling on a grpc thread to inspect the exception type and routes the record to either bad or failure bucket
- then the finishBundle thread waits for all of the futures to complete, ignores their value and uses the bad/failure bucket

I think it would be a lot better if you move the routing into finishBundle as well, this way processElement schedules async rps and stores the future. And finishBundle routes & emits things

Also, I have some concerns about large bundles causing OOMs and never emitting anything. I think it can be improved by having processElement poll the future state to see if anything is complete and process them prior to finishBundle

Reply:
This is already what is being done in the current flow. In the existing flow, writes and then exception handling happens on a thread spun up in processElement, and then all those threads are awaited in finishBundle. It happens under the hood a bit, and is called in bigtableWriter.close, which awaits on the bundler. I'll push the await into the bigtablewriter.close();

Reply:
I've pushed this wait down to bigtablewiter.close